### PR TITLE
Feature/cbv refactor

### PR DESCRIPTION
**Pull request by maraujop**
This is not a serious refactor, just a fast hack that works.

Using `render` method to redo flow execution makes the code hard to follow. It is necessary to read and understand django-formwizard flow to know that you need to call `get_prev_step` in that method. I would remove and change it for something like `process_flow_step` in which you can get actual current step calling `get_current_step`

Back button does not work, I find this more important than having a previous button somewhere in my form. I have redone process_post_request to use `form_prev_step` as a parameter that indicates the step form that called the wizard. This fixes the back button issue.

The code is not using class based views, sorry to use devel branch. I sent you a PM with some other stuff, I would like to contribute to the project.

Best regards,
Miguel Araujo

### CODE REVIEW

**Comment by stephrdev** [comment]
I updated my class based views refactor to make it work with the current Django trunk. Feel free to update your branch including your process_flow_step (and maybe explain the idea behind this method more detailed). Cheers, steph

**Comment by maraujop** [comment]
Happy new Year 2011 !

When I looked at your Django-app some weeks ago, I actually took some notes on things I wanted to comment with you, but I trashed them as I couldn't contact you, stupid thing, I know.
- From the top of my head. One thing that actually disturbed me, was the fact that django-formwizard didn't allow me to have several wizards start on the same view. Picture a URL where you have several forms that are wizard forms. If you are expecting a GET to initialize the wizard, you would need a wizard dispatcher or so, to be able to manage several forms. Not sure if I'm explaining myself. That's why I turned `form_prev_step` into a hidden field that explicitly says what step we are in and fixes the browser's back button issues.
- `get_form` method has a `step=None` parameter. step can be None and I think that should not happen. In my opinion this method should have a compulsory `step` parameter. Sometimes you don't pass a step parameter and others you do. Example:
  in `render_next_step` you pass `next_step`. So if for finding the current step someone is using `dertermine_step` it will fail, because you have not called `set_current_step` beforehand. That's why I say when you call `determine_step` it should return current step all times. 
- Another thing that confused me, is that you name some methods as named in `contrib.formtools` but they do different things.
- If I'm not mistaken `render_template` is recommended to be used to change the wizard flow in contrib. The problem is related to what `step` is within that method. `render_next_step` calls `render` which calls `render_template`. As  `storage.set_current_step(next_step)` is called before calling `render`, then if you do `storage.get_current_step` or `self.determine_step` you get the next step, not the current one. 

The reason I added `process_flow_step` is because I wanted to be able to call those methods(get_current_step or determine_step). I want to process the flow depending on the current step, not the next one. The other reason is that I like to overwrite methods and not having to call `super` on those methods, because I find that approach cleaner and less confusing, like in `get_template_context`

I hope in a couple days I can start writing a small draft for the documentation. I haven't tested your new CBV version, as I'm not using Django 1.3 yet.

Kind regards, 
Miguel

**Comment by stephrdev** [comment]
Hi Miguel,

I'm under heavy load at the moment. Sorry for taking some time to answer.

Regarding your "form_prev_step", idea. Why do you think this could break Go to previous step? The approach process_flow_step is fine with me. Please, use the current codebase and make your changes.

On Jan, 22. a sprint will occur in Berlin. I will review your stuff and hopefully pull it.

Cheers, steph

**Comment by maraujop** [comment]
Ok Steph,

I understand, no worries. 

Imagine someone is in a wizard created with django-formwizard and moves to step2, now presses back button in the browser. What happens in the new CBV version? At least in the old one, it breaks, because the storage is not reset as there is no way to know the user moved back. So I think having an explicit field that tells which step we are in, is crucial.

Now imagine you have a formwizard in your urls, but consider that you want to start 2 different wizards in the same URL, what do you do?

Did I explain myself this time better? 

I will get myself a virtual environment with Django 1.3 to test your app and work on it, but I'm also extremely busy right now, not sure how long it will take.

Cheers,
Miguel
